cast/java/ecj/src/main/java/com/ibm/wala/cast/java/translator/jdt/JDTJava2CAstTranslator.java

```diff
@@ -929,6 +943,127 @@ else if ((n.getModifiers() & Modifier.ABSTRACT) != 0) {
+      CAstType td = fTypeDict.getCAstTypeFor(tmp.resolveBinding().getType());
+      castTypes.add(td);
+      String tmpName = tmp.getName().getIdentifier();
+      castNames.add(tmpName);
+    }
+
+    // pass in memberEntities to the context, later visit(New) etc. may add classes
+    final Map<CAstNode, CAstEntity> memberEntities = HashMapFactory.make();
+    final MethodContext context =
+        new MethodContext(oldContext, memberEntities); // LEFTOUT: in polyglot there is a
+    // class context in between method and
+    // root
+
+    CAstNode mdast = fFactory.makeNode(CAstNode.RETURN, visitNode(n.getBody(), context));
+
+    // Polyglot comment: Presumably the MethodContext's parent is a ClassContext,
+    // and he has the list of initializers. Hopefully the following
+    // will glue that stuff in the right place in any constructor body.
+
+    if (context.getNameDecls() != null && !context.getNameDecls().isEmpty()) {
+      // new first statement will be a block declaring all names.
+      mdast =
+          fFactory.makeNode(
+              CAstNode.BLOCK_STMT,
+              context.getNameDecls().size() == 1
+                  ? context.getNameDecls().iterator().next()
+                  : fFactory.makeNode(CAstNode.BLOCK_STMT, context.getNameDecls()),
+              mdast);
+    }
+
+    /* TODO:
+     * createClassDeclaration creates a Function type when instantiating
+     * the lambda subclass. This is not correct, so a new type will have
+     * to be implemented.
+     * */
+
+    IMethodBinding fm = typeBinding.getFunctionalInterfaceMethod();
+    ITypeBinding ct = (fm == null ? n.resolveTypeBinding() : fm.getDeclaringClass());
+
+    CAstEntity lambdaClass =
+        createClassDeclaration(
+            n,
+            Collections.emptyList(),
+            null,
+            ct,
+            JDT2CAstUtils.anonTypeName(ct),
+            0 /* no modifiers */,
+            false,
+            false,
+            context,
+            null);
+
+    castNames.add(0, "this");
+    castTypes.add(0, lambdaClass.getType());
+
+    CAstEntity e =
+        new ProcedureEntity(
+            mdast,
+            ct,
+            Collections.emptyMap(),
+            context,
+            Collections.emptySet(),
+            castTypes,
+            castNames) {
+          private final CAstType myType =
+              new JdtMethodCAstType() {
+
+                @Override
+                public int getArgumentCount() {
+                  return castTypes.size() - 1;
+                }
+
+                @Override
+                public List<CAstType> getArgumentTypes() {
+                  return castTypes.stream()
+                      .skip(1)
+                      .map(
+                          x ->
+                              fm == null || fm.getName().equals("apply")
+                                  ? fTypeDict.getCAstTypeFor(
+                                      ast.resolveWellKnownType("java.lang.Object"))
```

Comment:
Why do we special-case `apply` here?

Reply:
this is because currently we erase types for Function but not for other types.

Reply:
> this is because currently we erase types for Function but not for other types.

Why?  And what if there is another functional interface whose method is named `apply`?